napari_animation/frame_sequence.py

```diff
@@ -0,0 +1,162 @@
+
+import numpy as np
+from napari.utils.events import EmitterGroup
+
+from .interpolation import Interpolation, InterpolationMap
+from .key_frame import ViewerState
+from .utils import pairwise
+
+if TYPE_CHECKING:
+    import napari
+
+    from .key_frame import KeyFrame, KeyFrameList
+
+
+class FrameSequence(Sequence[ViewerState]):
+    """Final sequence of of rendered animation frames, based on keyframes.
+
+    This object acts like an immutable sequence of frames, interpolated from
+    a sequence of (mutable) KeyFrames.  It can be indexed at any (valid) frame
+    in the animation, and will inteprolate (and cache) viewer state on demand.
+
+    If the KeyFrameList changes in any way, the cache is cleared.
+
+    Parameters
+    ----------
+    key_frames : KeyFrameList
+        A KeyFrameList from which to render the final frame sequence.
+    """
+
+    def __init__(self, key_frames: KeyFrameList) -> None:
+        super().__init__()
+        self._key_frames = key_frames
+        key_frames.events.inserted.connect(self._rebuild_frame_index)
+        key_frames.events.removed.connect(self._rebuild_frame_index)
+        key_frames.events.changed.connect(self._rebuild_frame_index)
+        key_frames.events.reordered.connect(self._rebuild_frame_index)
+
+        self.events = EmitterGroup(source=self, n_frames=None)
+
+        self.state_interpolation_map: InterpolationMap = {
+            "camera.angles": Interpolation.SLERP,
+            "camera.zoom": Interpolation.LOG,
+        }
+
+        # cache of interpolated viewer states
+        self._cache: Dict[int, ViewerState] = {}
+
+        # map of frame number -> (kf0, kf1, fraction)
+        self._frame_index: Dict[int, Tuple[KeyFrame, KeyFrame, float]] = {}
+        self._rebuild_frame_index()
+
+    def _rebuild_frame_index(self, event=None):
+        """Create a map of frame number -> (kf0, kf1, fraction)"""
+        self._frame_index.clear()
+        self._cache.clear()
+        if len(self._key_frames) < 2:
+            self.events.n_frames(value=len(self))
+            return
+
+        f = 0
+        for kf0, kf1 in pairwise(self._key_frames):
+            for s in range(kf1.steps):
+                fraction = s / kf1.steps
+                self._frame_index[f] = (kf0, kf1, fraction)
+                f += 1
+        self._frame_index[f] = (kf1, kf1, 0)
+        self.events.n_frames(value=len(self))
+
+    def __len__(self) -> int:
+        """The total frame count of the animation"""
+        return len(self._frame_index)
+
+    def __getitem__(self, key: int) -> ViewerState:
+        """Get the interpolated state at frame `key` in the animation."""
+        if key < 0:
+            key += len(self)
+        if key not in self._cache:
+            try:
+                kf0, kf1, frac = self._frame_index[key]
+            except KeyError:
+                raise IndexError(
+                    f"Frame index ({key}) out of range ({len(self)} frames)"
+                )
```

Comment:
getting errors here when playing with the `AnimationSlider` in the gui, will look into that

Reply:
all set